src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD201CancelAfterSwitchToMainThreadAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+
+namespace Microsoft.VisualStudio.Threading.Analyzers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using CodeAnalysis;
+    using CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Operations;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class VSTHRD201CancelAfterSwitchToMainThreadAnalyzer : DiagnosticAnalyzerBase
+    {
+        public const string Id = "VSTHRD201";
+
+        internal static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
+            Id,
+            title: Strings.VSTHRD201_Title,
+            messageFormat: Strings.VSTHRD201_MessageFormat,
+            helpLinkUri: Utils.GetHelpLink(Id),
+            category: "Style",
+            defaultSeverity: DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(startCtxt =>
+            {
+                var jtfSymbol = startCtxt.Compilation.GetTypeByMetadataName(Types.JoinableTaskFactory.FullName);
+                if (jtfSymbol != null)
+                {
+                    var switchMethods = jtfSymbol.GetMembers(Types.JoinableTaskFactory.SwitchToMainThreadAsync);
+                    var cancellationTokenTypeSymbol = startCtxt.Compilation.GetTypeByMetadataName(typeof(CancellationToken).FullName);
+                    startCtxt.RegisterSyntaxNodeAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, switchMethods, cancellationTokenTypeSymbol)), SyntaxKind.InvocationExpression);
+                }
+            });
+        }
+
+        internal static ExpressionSyntax GetCancellationTokenInInvocation(InvocationExpressionSyntax invocation, SemanticModel semanticModel, INamedTypeSymbol cancellationTokenTypeSymbol)
+        {
+            // Consider that named arguments allow for alternative ordering.
+            return invocation.ArgumentList.Arguments.FirstOrDefault(arg => semanticModel.GetTypeInfo(arg.Expression).Type?.Equals(cancellationTokenTypeSymbol) ?? false)?.Expression;
+        }
+
+        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context, ImmutableArray<ISymbol> switchMethods, INamedTypeSymbol cancellationTokenTypeSymbol)
+        {
+            var invocationSyntax = (InvocationExpressionSyntax)context.Node;
+            var invokeMethod = context.SemanticModel.GetSymbolInfo(context.Node).Symbol as IMethodSymbol;
```

Comment:
⚠️ `GetSymbolInfo` is costly in a syntax node analyzer. It's fine when used sparingly, but this one is not. This analyzer would benefit _substantially_ from switching to an `IOperation` analyzer. IOperation analyzers run as a group after the binding phase, and operate on a fully-cached semantic model of the code.

You also have access to things like `IArgumentOperation` which unambiguously match arguments to parameters so you don't need to worry about ordering of named arguments. Finally, IOperation analyzers are much easier to implement for both Visual Basic and C#.

Reply:
I recall trying the operation API for this analyzer initially. It was my first attempt, and I don't remember why I gave up. I'll try again and reach out if I need help. :)

Reply:
OK, so I have it written in terms of operations now, but I still have to call up the semantic model to get symbol and type symbol info. Is that a problem?

Also: to really get rid of my syntax-based search for the CancellationToken parameter, I have to pass an additionalLocation to my code fix. But that causes half of my tests to fail beacuse the analyzer test fx I'm using asserts that there are 0 additional locations. Can I just configure to omit that assertion since additional locations is an implementation detail of the contract I have between my analyzer and its code fix?

Reply:
When `IOperation` is used, it only is called on semantically correct code. So the symbols are already bound by the time they're queried, afaik, thus there's no worry about re-computing symbol information.

Reply:
> When IOperation is used, it only is called on semantically correct code.

I don't believe that's the case :)    If so, we would only be able to offer fixes if your code was already correct... thus defeating the purpose :)

>  So the symbols are already bound by the time they're queried, afaik, thus there's no worry about re-computing symbol information.

This part is true.  You can def "call up the semantic model to get symbol and type symbol info" when you have an IOp analyzer.  Note that IOps are file-based.  As such, they already expose `.SematicModel` directly off hte IOp.  YOu should def use that semantic model and not get a new one.

Reply:
> Note that IOps are file-based. As such, they already expose .SematicModel directly off hte IOp

`OperationAnalysisContext.Compilation.GetSemanticModel` is what I'm using. I don't see a more "local" semantic model off of `IOperation` directly to use. Am I missing something?